File: myFinance_improved_V2/lib/features/sales_invoice/data/models/cash_location_model.dart
import '../../domain/entities/cash_location.dart';

/// Cash location model for data transfer
///
/// Supports get_cash_locations_v2 RPC response format:
/// {
///   "cash_location_id": "uuid",
///   "location_name": "Location Name",
///   "location_type": "cash" | "bank" | "vault",
///   "store_id": "uuid" | null,
///   "store_name": "Store Name" | null,
///   "is_company_wide": true | false,
///   "currency_code": "VND" | null,
///   "bank_account": "..." | null,
///   "bank_name": "..." | null,
///   ...
/// }
class CashLocationModel {
  final Map<String, dynamic> json;

  CashLocationModel(this.json);

  /// Create from JSON
  factory CashLocationModel.fromJson(Map<String, dynamic> json) {
    return CashLocationModel(json);
  }

  /// Convert to domain entity (supports both v1 and v2 response formats)
  CashLocation toEntity() {
    // v2 uses snake_case, v1 uses camelCase
    return CashLocation(
      id: json['cash_location_id']?.toString() ?? json['id']?.toString() ?? '',
      name: json['location_name']?.toString() ?? json['name']?.toString() ?? '',
      type: json['location_type']?.toString() ?? json['type']?.toString() ?? 'cash',
      storeId: json['store_id']?.toString() ?? json['storeId']?.toString() ?? '',
      isCompanyWide: (json['is_company_wide'] as bool?) ?? (json['isCompanyWide'] as bool?) ?? false,
      currencyCode: json['currency_code']?.toString() ?? json['currencyCode']?.toString() ?? 'VND',
      bankAccount: json['bank_account']?.toString() ?? json['bankAccount']?.toString(),
      bankName: json['bank_name']?.toString() ?? json['bankName']?.toString(),
    );
  }
}

File: website/src/features/sale_product/data/models/CashLocationModel.ts
/**
 * CashLocationModel
 * DTO and Mapper for CashLocation entity
 * Updated for get_cash_locations_v2 RPC response
 */

import { CashLocation, CashLocationType } from '../../domain/entities/CashLocation';

/**
 * DTO matching get_cash_locations_v2 RPC response
 */
export interface CashLocationDTO {
  cash_location_id: string;
  location_name: string;
  location_type: CashLocationType;
  store_id: string | null;
  store_name: string | null;
  company_id: string;
  is_company_wide: boolean;
  currency_code: string | null;
  currency_id: string | null;
  bank_account: string | null;
  bank_name: string | null;
  location_info: string | null;
  icon: string | null;
  note: string | null;
  main_cash_location: boolean;
  created_at: string;
  created_at_utc: string | null;
}

export class CashLocationModel {
  /**
   * Convert DTO to Domain Entity
   */
  static toDomain(dto: CashLocationDTO): CashLocation {
    return CashLocation.create({
      id: dto.cash_location_id,
      name: dto.location_name,
      type: dto.location_type,
      storeId: dto.store_id,
      isCompanyWide: dto.is_company_wide,
      currencyCode: dto.currency_code,
      bankAccount: dto.bank_account,
      bankName: dto.bank_name,
    });
  }

  /**
   * Convert Domain Entity to DTO
   */
  static fromDomain(cashLocation: CashLocation): Partial<CashLocationDTO> {
    return {
      cash_location_id: cashLocation.id,
      location_name: cashLocation.name,
      location_type: cashLocation.type,
      store_id: cashLocation.storeId,
      is_company_wide: cashLocation.isCompanyWide,
      currency_code: cashLocation.currencyCode,
      bank_account: cashLocation.bankAccount,
      bank_name: cashLocation.bankName,
    };
  }

  /**
   * Convert array of DTOs to Domain Entities
   */
  static toDomainList(dtos: CashLocationDTO[]): CashLocation[] {
    return dtos.map((dto) => CashLocationModel.toDomain(dto));
  }

  /**
   * Convert to Domain (v2 doesn't return deleted items)
   */
  static filterAndConvertActive(dtos: CashLocationDTO[]): CashLocation[] {
    return dtos.map((dto) => CashLocationModel.toDomain(dto));
  }
}
